### ISSUE TITLE
Addition mechanism for custom post-processing

### ISSUE THREAD

**Issue by harrison314** [ZarehD/AspNetStatic#17]
**Is your feature request related to a problem? Please describe.**
I found that the HTML minifier is unable to remove Blazor SSR HTML comments. This led me to the idea that it would be a good idea to allow custom post-processing of the resulting files.

This can be used for resource compression, image scaling, webp conversion, uploading the resulting page to a server, or creating a zip package for deployment...

It would give the user a free hand.

**Describe the solution you'd like**
As a solution, I considered adding an eg. `IPostProcessor` interface, which an AspNetStatic user can implement and add to DI.

```cs
public interface IPostProcessor
{
   void Initialize(string destinationRoot); // call once before processing individual files
   void PostProcess(ResourceInfoBase resource, string resourceDestinationPath);  //call for all resources files
   void Finalize(string destinationRoot);  // call once after processing individual files
}
```

In `app.GenerateStaticContent(...)` an `IEnumerable<IPostProcessor>` would be obtained from DI and then executed on all created files. This does not force the API user to implement `IPostProcessor`, but they can implement multiple ones at the same project.

The method and interface names are just an example.

**Comment by ZarehD** [comment]
Hello @harrison314, and thank you for the suggestion. Yes, I'm open to the idea of providing a way for a user to define pre/post-processing operations to be performed on each resource as it is processed. Just to confirm, though, something like the following flow is what you are suggesting:
```C#
// IHtmlPrePostProcessor, ICssPrePostProcessor, IJsPrePostProcessor, IBinPrePostProcessor
var userDefPageProcessor = services.GetService<IHtmlPrePostProcessor>();
var doUserProcessing = userDefPageProcessor is not null;
var shouldOptimizeContent = userDefPageProcessor?.AllowDefaultOptimization ?? true;

foreach (var p in pageResources)
{
    string content = GetPageResourceContent(p);
    if (doUserProcessing)
    {
        content = userDefPageProcessor.PreProcess(p, content);
    }
    content = ProcessPageContent(p, content);
    if (doUserProcessing)
    {
        content = userDefinedHtmlProcessor.PostProcess(p, content);
    }
    if (shouldOptimizeContent)
    {
        content = OptimizePageContent(p, content);
    }
    SaveStaticHtmlToFile(p, content);
}
```
I will take a closer look at the implementation options and work this into my schedule at the earliest opportunity. Thanks again.

**Comment by harrison314** [comment]
Hello,
yes, the above processing method also evaluates my requirements.

Thanks.

**Comment by ZarehD** [comment]
Hello @harrison314,

As I look more closely at this request, it appears that most of what's needed already exists; it simply needs to be utilized a/o made a bit simpler.

I'm referring to the `IOptimizerSelector` interface, of course. While the interface is used by the existing implementation (`OptimizerSelector`), the service is not requested/injected when it's needed. I'll be rectifying that!

In addition, the optimizer interfaces could use some work. Right now, I'm just using the WebMarkupMin interfaces (mostly). I will change this to use more agnostic interfaces (i.e. `IMarkupOptimizer`, `ICssOptimizer`, etc.)

So, the upshot is that by registering a custom `IOptimizerSelector` implementation, you will be able to select your own custom optimizer to call for a given resource type. The custom "optimizer" implementation could in turn use one of the existing optimizers (e.g. `WebMarkupMin` minifier) to do the basic work before/after doing its own specialized work.

JFC, an "optimizer" can do any sort of processing on the content. The in-box optimizers use `WebMarkupMin` to minify the content, but you can do anything you want (including incorporating the `WebMarkupMin` functionality into your own custom steps).

I'm looking at knocking this out in the next few days, so please let me know if you have any thoughts on this. Thanks.

**Comment by harrison314** [comment]
Hello,

> So, the upshot is that by registering a custom IOptimizerSelector implementation, you will be able to select your own custom optimizer to call for a given resource type. The custom "optimizer" implementation could in turn use one of the existing optimizers (e.g. WebMarkupMin minifier) to do the basic work before/after doing its own specialized work.

With such a solution, I am afraid that as a user of the library I will have to use "internal" library classes that are registered in DI. And this makes future changes difficult for the library creator.

Won't this also change the function of the switches in the `GenerateStaticContent` method?

**Comment by ZarehD** [comment]
The optimizers have (nor need) any access to the library's "internals". They get handed "content" to process however they wish before returning it. As a user of the library, it is intended for you to be able to customize this functionality by registering your own implementation(s). This is pretty standard stuff. To implement one of the interfaces, subclass & override the default implementation, or code it from scratch if that's better for your needs. I'd be glad to provide more guidance if you need it.

WRT flags/switches, I don't see any issue at all. Is there something specific you can point to?

**Comment by harrison314** [comment]
>The optimizers have (nor need) any access to the library's "internals". They get handed "content" to process however they wish before returning it. As a user of the library, it is intended for you to be able to customize this functionality by registering your own implementation(s). This is pretty standard stuff. To implement one of the interfaces, subclass & override the default implementation, or code it from scratch if that's better for your needs. I'd be glad to provide more guidance if you need it.

In that case, if it is documented which class needs to be inherited or which interface needs to be implemented, then I am OK with that.

> WRT flags/switches, I don't see any issue at all. Is there something specific you can point to?

I meant this `app.GenerateStaticContent(dontOptimizeContent: false)`, but now that I look at it, it's not a problem, because `dontOptimizeContent` turns off all optimization and not just the one that would be the default.

**Comment by ZarehD** [comment]
Yes, of course, I will definitely document which interfaces you'd need to implement (or classes to inherit from).

**Comment by ZarehD** [comment]
The relevant updates have been published so I'm closing this ticket, but please let me know if you run into any problems.

**Comment by harrison314** [comment]
I tried it and I encountered two problems.


1. `MarkupOptimizerResult.OptimizedContent` is readonly, which is contrary to the documentation.
```cs
internal partial class ContentOptimizer : IMarkupOptimizer
{
    private readonly DefaultMarkupOptimizer defaultMarkupOptimizer;

    [GeneratedRegex("<!--Blazor.*?-->\r?\n?", RegexOptions.Singleline)]
    private static partial Regex GetRemoveCommentRegex();

    public ContentOptimizer(DefaultMarkupOptimizer defaultMarkupOptimizer)
    {
        this.defaultMarkupOptimizer = defaultMarkupOptimizer;
    }


    public MarkupOptimizerResult Execute(string content, PageResource resource, string outFilePathname)
    {
        MarkupOptimizerResult result = this.defaultMarkupOptimizer.Execute(content, resource, outFilePathname);
        result.OptimizedContent = this.ExecuteCustomOptimization(result.OptimizedContent);

        return result;
    }

    private string ExecuteCustomOptimization(string content)
    {
        content = GetRemoveCommentRegex().Replace(content, string.Empty);

        return content;
    }
}
```

2. Error in DI

```cs
        builder.Services.AddSingleton<IMarkupOptimizer, ContentOptimizer>();
        builder.Services.AddDefaultOptimizers();
```

```
System.AggregateException: 'Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: AspNetStatic.Optimizer.IMarkupOptimizer Lifetime: Singleton ImplementationType: Harrison314Blog.ContentOptimizer': Unable to resolve service for type 'WebMarkupMin.Core.XmlMinificationSettings' while attempting to activate 'AspNetStatic.Optimizer.DefaultMarkupOptimizer'.) (Error while validating the service descriptor 'ServiceType: AspNetStatic.Optimizer.DefaultMarkupOptimizer Lifetime: Singleton ImplementationType: AspNetStatic.Optimizer.DefaultMarkupOptimizer': Unable to resolve service for type 'WebMarkupMin.Core.XmlMinificationSettings' while attempting to activate 'AspNetStatic.Optimizer.DefaultMarkupOptimizer'.) (Error while validating the service descriptor 'ServiceType: AspNetStatic.Optimizer.IMarkupOptimizer Lifetime: Singleton ImplementationType: AspNetStatic.Optimizer.DefaultMarkupOptimizer': Unable to resolve service for type 'WebMarkupMin.Core.XmlMinificationSettings' while attempting to activate 'AspNetStatic.Optimizer.DefaultMarkupOptimizer'.)'
```

Branch from code is here https://github.com/harrison314/harrison314.github.io.src/tree/TryCustomOptimization

**Comment by ZarehD** [comment]
Yes, I see. The `OptimizedContent` property was meant to only be set by the ctor. But the use-case I showed acts as tho it's read/write. So be it. I will make it read/write.

The second error is a bit more interesting given that that all `DefaultMarkupOptimizer` ctor parms are nullable (i.e. not required). Lets try this; please put the `AddDefaultOptimizers` (or `AddDefaultMinifier`) call before your `AddSingleton()` call.

**Comment by ZarehD** [comment]
Just pushed v0.26.4 with the necessary fixes. Thanks for the feedback, @harrison314.

**Comment by harrison314** [comment]
@ZarehD Thanks. The two problems mentioned above have been solved.

But my `IMarkupOptimizer` was never called.


MarkupOptimizer: <https://github.com/harrison314/harrison314.github.io.src/blob/TryCustomOptimization/src/Harrison314Blog/Harrison314Blog/ContentOptimizer.cs#L19>

Registration: <https://github.com/harrison314/harrison314.github.io.src/blob/TryCustomOptimization/src/Harrison314Blog/Harrison314Blog/Program.cs#L39>

**Comment by ZarehD** [comment]
The order of registration calls is important. It's basically a last-in-wins model; the last IMarkupOptimizer to be registered is the one that wins. By calling `AddDefaultOptimizers()` _after_ registering your custom optimizer you're replacing it with the default optimizer. So, just move the `AddDefaultOptimizers` call further up (earlier) in your code, above/before where you register your custom optimizer(s).

**Comment by harrison314** [comment]
Thank you, it helped. It is still necessary to update the documentation - first the custom optimizer is registered and then the default one.

See: https://github.com/ZarehD/AspNetStatic/blob/master/Readme.md?plain=1#L454

**Comment by ZarehD** [comment]
The docs actually show the `AddDefaultOptimizers` call in relation to registering a custom `IOptimizerSelector`, but it's not a bad idea to add something related to custom optimizers too. Thanks.
